cpp/src/parquet/encoding.cc

```diff
@@ -2250,17 +2269,17 @@ void DeltaBitPackEncoder<DType>::FlushBlock() {
         std::min(values_per_mini_block_, values_current_block_);
 
     const uint32_t start = i * values_per_mini_block_;
-    const UT max_delta = *std::max_element(
+    const T max_delta = *std::max_element(
         deltas_.begin() + start, deltas_.begin() + start + values_current_mini_block);
 
     // The minimum number of bits required to write any of values in deltas_ vector.
     // See overflow comment above.
-    const auto bit_width = bit_width_data[i] =
-        bit_util::NumRequiredBits(max_delta - min_delta);
+    const auto bit_width = bit_width_data[i] = bit_util::NumRequiredBits(
+        static_cast<UT>(SafeSignedSubtractSigned(max_delta, min_delta)));
```

Comment:
Actually I'm not fully understand here. If the sequence is `{1, 0, -1, 0, 1, 0, -1, 0, 1}`

Previously, the value glows large because `max_delta - min_delta`. The `max_delta` should be 1, however, it turns to be a huge unsigned int, and the `min_delta` is 1. The result is correct, however it waste huge space?

Reply:
Yes, you are correct. Leaving everything unsigned, we have (unsigned)(0 - 1) = 0xffffffffU. `min_delta == 1`, `max_delta == 0xffffffffU`, so `max_delta - min_delta == 0xfffffffeU`, which requires the full 32 bits to encode.